wire-ios-data-model/Source/Model/Conversation/ZMConversation+Predicates.swift

```diff
@@ -77,102 +77,124 @@ extension ZMConversation {
+
+        return NSCompoundPredicate(andPredicateWithSubpredicates: [isTeamConversation, isGroupConversation, hasNoUserDefinedName, hasOnlyOneParticipant])
     }
 
-    @objc(predicateForClearedConversations)
-    class func predicateForClearedConversations() -> NSPredicate {
-        let cleared = NSPredicate(format: "\(ZMConversationClearedTimeStampKey) != NULL AND \(ZMConversationIsArchivedKey) == YES")
+    class func predicateForConversationsNeedingToBeCalculatedUnreadMessages() -> NSPredicate {
+         return NSPredicate(format: "%K == YES", ZMConversationNeedsToCalculateUnreadMessagesKey)
+    }
 
-        return NSCompoundPredicate(andPredicateWithSubpredicates: [cleared, predicateForValidConversations()])
+    public static func predicateForConversationsArePendingToRefreshMetadata() -> NSPredicate {
+        return NSPredicate(format: "\(ZMConversationIsPendingMetadataRefreshKey) == YES")
     }
 
-    @objc(predicateForConversationsIncludingArchived)
-    class func predicateForConversationsIncludingArchived() -> NSPredicate {
-        let notClearedTimestamp = NSPredicate(format: "\(ZMConversationClearedTimeStampKey) == NULL OR \(ZMConversationLastServerTimeStampKey) > \(ZMConversationClearedTimeStampKey) OR (\(ZMConversationLastServerTimeStampKey) == \(ZMConversationClearedTimeStampKey) AND \(ZMConversationIsArchivedKey) == NO)")
+}
 
-        return NSCompoundPredicate(andPredicateWithSubpredicates: [notClearedTimestamp, predicateForValidConversations()])
-    }
+extension String {
 
-    @objc(predicateForGroupConversations)
-    class func predicateForGroupConversations() -> NSPredicate {
-        let groupConversationPredicate = NSPredicate(format: "\(ZMConversationConversationTypeKey) == \(ZMConversationType.group.rawValue)")
-        let notInFolderPredicate = NSCompoundPredicate(notPredicateWithSubpredicate: predicateForConversationsInFolders())
+    var words: [String] {
+        var words: [String] = []
+        enumerateSubstrings(in: self.startIndex..., options: .byWords) { substring, _, _, _ in
+            words.append(String(substring!))
+        }
 
-        return .all(of: [
-            predicateForConversationsExcludingArchived(),
-            groupConversationPredicate,
-            notInFolderPredicate
-        ])
+        if words.isEmpty {
+            words = [trimmingCharacters(in: .whitespacesAndNewlines)]
+        }
+
+        return words
     }
+}
 
-    @objc(predicateForLabeledConversations:)
-    class func predicateForLabeledConversations(_ label: Label) -> NSPredicate {
-        let labelPredicate = NSPredicate(format: "%@ IN \(ZMConversationLabelsKey)", label)
+@objc
+public final class ConversationPredicateFactory: NSObject {
 
-        return NSCompoundPredicate(andPredicateWithSubpredicates: [predicateForConversationsExcludingArchived(), labelPredicate])
+    private let selfTeam: Team?
+
+    @objc
+    init(selfTeam: Team? = nil) {
```

Comment:
I ask myself why we need retain team as object here. Is it required for all funcs or only certain ones? In my opinion a factory should not depend on any state but provide pure functions.

Reply:
It's used in pretty much all of the predicate functions because it's used in one of the lowest level predicates (the one that filters out valid one on one conversations). I chose to retain it here to avoid injected it in every function of the factory, but I'm happy to reconsider. I would though defer that for a follow up PR that could simplify these predicates.

Reply:
Okay fine